walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,140 @@
+      if (event.getAttendees().contains(attendee)) {
+        int eventStart = event.getWhen().start();
+        int eventEnd = event.getWhen().end();
+
+        availableTimes.add(TimeRange.fromStartEnd(availableStart, eventStart, false));
+        availableStart = eventEnd;
+      }
+    }
+    availableTimes.add(TimeRange.fromStartEnd(availableStart, TimeRange.END_OF_DAY, true));
+
+    return availableTimes;
+  }
+
+  /**
+   * Given a {@code TimeRange} list of lists, return the interval list intersections.
+   *
+   * @param allLists The {@code TimeRange} list of lists that we wish to get the intersection of.
+   * @return The intersection of all the lists given.
+   */
+  private ArrayList<TimeRange> allTimeRangesIntersection(ArrayList<ArrayList<TimeRange>> allLists) {
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    // Whole day TimeRange is a base case because it is the identity element, and should also be
+    // returned when no attendees are given.
+    availableTimes.add(TimeRange.WHOLE_DAY);
+    for (ArrayList<TimeRange> ranges : allLists) {
+      availableTimes = twoTimeRangesIntersection(availableTimes, ranges);
+    }
+    return availableTimes;
+  }
+
+  /**
+   * Given two {@code TimeRange} ArrayLists, returns the intersection of the two interval lists.
+   *
+   * Example: |---| |--|
+   *            |-----|
+   * Returns:   |-| |-|
+   *
+   * @param arr1 The first ArrayList to intersect with.
+   * @param arr2 The second ArrayList to intersect with.
+   * @return The intersection of the two ArrayLists.
+   */
+  private ArrayList<TimeRange> twoTimeRangesIntersection(
+      ArrayList<TimeRange> arr1, ArrayList<TimeRange> arr2) {
+    ArrayList<TimeRange> intersection = new ArrayList<TimeRange>();
+    int arr1Index = 0, arr2Index = 0;
+
+    // Iterating through the TimeRange endpoints of the ArrayLists, add the overlap between the
+    // segments to the intersection.
+    while (arr1Index < arr1.size() && arr2Index < arr2.size()) {
+      TimeRange rangeFrom1 = arr1.get(arr1Index);
+      TimeRange rangeFrom2 = arr2.get(arr2Index);
+
+      if (rangeFrom1.overlaps(rangeFrom2)) {
+        int start = Math.max(rangeFrom1.start(), rangeFrom2.start());
+        int end = Math.min(rangeFrom1.end(), rangeFrom2.end());
+        intersection.add(TimeRange.fromStartEnd(start, end, false));
+      }
+
+      if (rangeFrom1.end() < rangeFrom2.end()) {
+        arr1Index++;
+      } else {
+        arr2Index++;
+      }
+    }
+
+    return intersection;
+  }
+
+  /**
+   * Removes the times in the given {@code ranges} parameter that is below the given time duration.
+   *
+   * @param ranges The list of {@code TimeRange}s to consider.
+   * @param duration The duration that all {@code TimeRange}s must not fall under.
+   */
+  private ArrayList<TimeRange> removeTimesBelowDuration(
+      ArrayList<TimeRange> ranges, long duration) {
+    ArrayList<TimeRange> newTimeRanges = new ArrayList<TimeRange>();
```

Comment:
Is there a more.. intuitive name for this ArrayList?

Reply:
Changed the variable name to "timeRangesAboveDuration"